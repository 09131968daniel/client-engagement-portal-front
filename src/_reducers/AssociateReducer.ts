export interface IAssociate {
    firstName?: string,
    lastName?: string,
    testScores?: {
        week: number,
        score: number
    }[],
    techScores?: {
        tech: string,
        score: number
    }[]
}
// got it. you pulled his code
export interface IAssociatesState{
    associates: IAssociate[];
}

//will set it up to hold all associates
const initialState: IAssociatesState = {
    associates: [],
};

// export const associateReducer = (state = initialState, action:{type:string, payload:IAssociate}): IAssociatesState => {
//     switch(action.type) {
//         case associateTypes.GET_ALL_ASSOCIATES:
//             return {
//                     associates: [...state.associates, action.payload],
//             };
//     }
//     return state;
// } 