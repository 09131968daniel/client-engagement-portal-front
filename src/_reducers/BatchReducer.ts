import { SET_BATCHES } from "../actions/BatchCardActions"

//Batch state is essentially just an array of batch objects
export interface IBatchState{
    batches: {
        id: number,
        skill: string,
        name: string
    }[];
}

export const initialBatchState:IBatchState = {
    batches: [{id: 0, skill: "", name: ""}] //empty array to start with
}

/**
 *
 * This reducer goes through every action type that exists in this program but will
 *  only respond to the one that's called SET_BATCHES
 *
 * @param batchState the batch card state in the store is passed in automatically
 * @param action goes through every action in the program
 *
 * @returns Batch state. Basically just updates the batches state
 */
export const BatchReducer = (batchState:IBatchState = initialBatchState, action:any) => {
    switch(action.type)
    {
        case SET_BATCHES:
            if(action.payload != null) {
                return { ...action.payload }  //places batch data into batchCard state
            }
            else {
                return batchState;
            }
        default:
            return batchState;
    }
}
