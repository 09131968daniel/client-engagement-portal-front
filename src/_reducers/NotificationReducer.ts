//basic structure of the notifications
export interface INotification {
    clientName:string
    subject:string
    requestDate:string
}

//state of the notifications
export interface INotificationState {
    notifications:INotification[] | null
}

//basic initial state just to display info
const initialState:INotificationState = {
        notifications: [
        {
            clientName:'a1',
            subject:'a2',
            requestDate:'a3'
        },
        {
            clientName:'b1',
            subject:'b2',
            requestDate:'b3'
        },
        {
            clientName:'c1',
            subject:'c2',
            requestDate:'c3'
        },
    ]
}

export const notificationReducer = 
    (state:INotificationState = initialState, action:{type:string, payload:INotification}):INotificationState => {
        switch(action.type) {
            case "GET_NOTIFICATIONS":

                //TODO: make an axios get request here to populate the notification state

                return {notifications: null};
            case "ADD_NOTIFICATION":
                if(state == null)
                    return {notifications: null};

                state = {
                    notifications:[
                        ...state.notifications as INotification[],
                        action.payload
                    ]
                };

                //TODO: make an axios post request here to add the notification

                return state;
            case "REMOVE_NOTIFICATION":
                if(state == null)
                    return {notifications: null};

                const payload = (element:INotification) => element = action.payload;

                const index = state.notifications?.findIndex(payload);

                state = {
                    notifications: state.notifications?.splice(index as number, 1) as INotification[]
                }

                //TODO: make an axios delete request here to remove the notification

                return state;
            default:
                return state;
        }
}