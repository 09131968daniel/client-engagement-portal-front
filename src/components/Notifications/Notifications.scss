$orange: #F26925;
$white:  #FFFFFF;
$darkgray:  #474C55;

#notif-header {
    background-color: $orange;
    color: $white;
}

#notif-header:hover {
    background-color: $darkgray;
}

.myTable {
    margin-bottom: 0px !important;
}

// .notif-row {
//     background-color: $darkgray;
//     color: $white;
// }