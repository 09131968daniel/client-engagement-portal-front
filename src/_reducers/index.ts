import { AnyAction, CombinedState, combineReducers, Reducer } from "redux";
import { BatchReducer, IBatchState } from "./BatchReducer";
import { IUserState, userReducer } from "./UserReducer";

export interface IRootState{
    userState:IUserState,
    batchState:IBatchState
}

export const rootReducer:Reducer<CombinedState<IRootState>, AnyAction> = combineReducers<IRootState>(
    {userState: userReducer,
    batchState:BatchReducer,
    //place your state:reducer here
});
