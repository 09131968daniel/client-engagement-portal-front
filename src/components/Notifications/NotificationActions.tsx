import { INotification } from "../../_reducers/NotificationReducer";

export const setNotifications = (notifs:INotification[]) => {
    return {
        type: "SET_NOTIFICATIONS",
        payload: notifs
    }
};
