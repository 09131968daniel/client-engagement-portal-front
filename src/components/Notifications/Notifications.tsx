import React from 'react';
import { Accordion, Card } from 'react-bootstrap';
import { Table } from 'reactstrap';
import { INotificationState } from '../../_reducers/NotificationReducer';
import './Notifications.scss';

//TODO: Add actions, styling

export const Notifications:React.FC<INotificationState> = (props:INotificationState) => {

    return (
        <Accordion defaultActiveKey="0">
                <Accordion.Toggle as={Card.Header} variant="link" eventKey="0" id="notif-header">
                    Notifications
                </Accordion.Toggle>
                
                {props.notifications?.map(
                        (e, i) =>
                        <Accordion.Collapse eventKey="0" key={i}>
                            <Table className="myTable">
                                <tbody>
                                    <tr className="notif-row">
                                        <td>{e.clientName}</td>
                                        <td>{e.subject}</td>
                                        <td>{e.requestDate}</td>
                                    </tr>
                                </tbody>
                            </Table>
                        </Accordion.Collapse>
                    )
                }
        </Accordion>
    )
}