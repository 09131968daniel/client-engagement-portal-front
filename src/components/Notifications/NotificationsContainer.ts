import { connect } from "react-redux";
import { IRootState } from "../../_reducers";
import { INotification } from "../../_reducers/NotificationReducer";
import { Notifications } from "./Notifications";

const mapStateToProps = (state:IRootState) => {
    return {
        notifications: state.notificationState.notifications
    }
}

export interface INotificationActions {
    addNotification:(cName:string, subj:string, date:string) => {type:string, payload:INotification}
    getNotifications:(cName:string, subj:string, date:string) => {type:string, payload:INotification}
    removeNotification:(cName:string, subj:string, date:string) => {type:string, payload:INotification}
}

const mapDispatchToProps:INotificationActions = {
    /**
     * Creates an ADD_NOTIFICATION action
     */
    addNotification:(cName:string, subj:string, date:string) => {
        return {
            type: "ADD_NOTIFICATION",
            payload: {
                clientName: cName,
                subject: subj,
                requestDate: date
            }
        }
    },

    /**
     * Creates a GET_NOTIFICATIONS action
     */
    getNotifications:(cName:string, subj:string, date:string) => {
        return {
            type: "GET_NOTIFICATIONS",
            payload: {
                clientName: cName,
                subject: subj,
                requestDate: date
            }
        }
    },

    /**
     * Creates a REMOVE_NOTIFICATION action
     */
    removeNotification:(cName:string, subj:string, date:string) => {
        return {
            type: "REMOVE_NOTIFICATION",
            payload: {
                clientName: cName,
                subject: subj,
                requestDate: date
            }
        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Notifications)