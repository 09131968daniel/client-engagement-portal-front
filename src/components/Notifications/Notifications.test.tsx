import configureStore from 'redux-mock-store'
import { notificationReducer } from '../../_reducers/NotificationReducer';
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux';
import Notifications from './Notifications';
import React from 'react';
import { addNotification, getNotifications, removeNotification } from './NotificationActions';

/**
 * Tests the notificationReducer and most of its functions
 */
describe('notification reducer', () => {
    /**
     * Tests the initial state by sending an action without a specific type.
     * The return statement should be an empty notifications array.
     */
    it('should return the initial state', () => {
        expect(notificationReducer(undefined, {type:"", payload:{clientName:"", subject: "", requestDate: ""}})).toEqual({notifications:[]});
    })

    /**
     * Tests the ADD_NOTIFICATION action by passing in an object and checking the state to confirm that it now has the object.
     * The return statement should be a notifications array with the payload added.
     */
    it('should handle ADD_NOTIFICATION', () => {
        expect(
            notificationReducer(undefined, {
                type: "ADD_NOTIFICATION",
                payload: {
                    clientName: "a1",
                    subject: "a2",
                    requestDate: "a3"
                }
            })
        ).toEqual({
            notifications: [
                {
                    clientName: "a1",
                    subject: "a2",
                    requestDate: "a3"
                }
            ]
        })
    })

    /**
     * Tests the REMOVE_NOTIFICATION action by having a default state with an object and then attempted to remove it from the state.
     * The return statement should be an empty notifications array.
     */
    it('should handle REMOVE_NOTIFICATION', () => {
        expect(
            notificationReducer({
                notifications: [ {
                    clientName: "a1",
                    subject: "a2",
                    requestDate: "a3"
                }]
            }, 
            {
                type: "REMOVE_NOTIFICATION",
                payload: {
                    clientName: "a1",
                    subject: "a2",
                    requestDate: "a3"
                }
            })
        ).toEqual({
            notifications: []
        })
    })
})

/**
 * Tests the notification actions
 */
describe('notification actions', () => {
    it('should return an add notification action', () => {
        expect(
            addNotification("a", "b", "c")
        ).toEqual(
            {
                type: "ADD_NOTIFICATION",
                payload: {
                    clientName: "a",
                    subject: "b",
                    requestDate: "c"
                }
            }
        )
    });

    it('should return a get notifications action', () => {
        expect(
            getNotifications("a", "b", "c")
        ).toEqual(
            {
                type: "GET_NOTIFICATIONS",
                payload: {
                    clientName: "a",
                    subject: "b",
                    requestDate: "c"
                }
            }
        )
    })

    it('should return a remove notification action', () => {
        expect(
            removeNotification("a", "b", "c")
        ).toEqual(
            {
                type: "REMOVE_NOTIFICATION",
                payload: {
                    clientName: "a",
                    subject: "b",
                    requestDate: "c"
                }
            }
        )
    })
})

// Tests that given a state, the notifications will render properly
describe('notification component', () => {

    const mockStore = configureStore([]);

    let store:any;
    let component:any;

    beforeEach(() => {
        store = mockStore({
            notificationState: {
                notifications: [
                    {
                        clientName: "a1",
                        subject: "a2",
                        requestDate: "a3"
                    }
                ]
            }
        });

        component = renderer.create(
            <Provider store={store}>
                <Notifications/>
            </Provider>
        )
    });

    it('should render the component with the given state', () => {
        expect(component.toJSON()).toMatchSnapshot();
    })
})
