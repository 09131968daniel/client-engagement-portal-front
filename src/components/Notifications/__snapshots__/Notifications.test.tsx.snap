// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notification component should render the component with the given state 1`] = `
<div
  className="notifs-container accordion"
>
  <div
    className="card-header"
    id="notif-header"
    onClick={[Function]}
    variant="link"
  >
    Notifications
  </div>
  <div
    aria-expanded={null}
    className="collapse"
  >
    <div>
      <div
        className="notifs accordion"
      />
    </div>
  </div>
</div>
`;
